### DISCUSSION TITLE
Equivalent to MrBayes's Symdirihyperpr prior

### DISCUSSION THREAD

**Discussion by ms609** [revbayes/revbayes#610] [Q&A] [answered]
MrBayes has an option `prset symdirihyperpr`, which allows for non-uniform stationary distributions of state frequencies, explored by Wright et al. ([Systematic Biology 2016]( https://doi-org.ezphost.dur.ac.uk/10.1093/sysbio/syv122)).

The [MrBayes manual](https://gensoft.pasteur.fr/docs/mrbayes/3.2.7/Manual_MrBayes_v3.2.pdf) explains:

> Because state labels are arbitrary in the standard model, we cannot estimate
unequal stationary state frequencies or substitution rates (recall that the
stationary state frequencies are an important factor in determining the latter).
However, it is still possible to allow the state frequencies (rates) to vary over sites
according to some appropriate distribution. MrBayes uses a symmetric Dirichlet
distribution for this purpose. For binary data, the analogy of the Dirichlet
distribution is called the beta distribution; MrBayes uses Dirichlet and beta
interchangeably for the distribution depending on context. The approach is
similar to the one used to allow rate variation across sites according to a gamma
distribution: we calculate the likelihood of a site assuming different discrete
categories of asymmetry and then we sum the values to obtain the site likelihood.
>
> The symmetric Dirichlet distribution has one parameter that determines its
shape, just like the alpha parameter determines the shape of the gamma
distribution. The larger the parameter of the symmetric Dirichlet, the less
transition rate (stationary frequency) asymmetry there is across sites. By default,
the parameter is fixed to infinity (prset symdirihyperpr=fixed(infinity) );
this corresponds to the standard assumption of no transition rate asymmetry
across sites: the rate of going from 0 to 1 is equal to the rate of going from 1 to 0
for all characters. The prior is called a hyperprior because it concerns a
distribution that controls the distributions of other model parameters (stationary
state frequencies in this case). If you want to allow transition rate (stationary
frequency) asymmetry in standard data, then simply select another (hyper)prior. For instance, you can fix the parameter to 1.0, which would result in a uniform
prior on the proportions of the state frequencies.

[`dnPhyloCTMC()`](https://revbayes.github.io/documentation/dnPhyloCTMC.html) doesn't obviously exhibit an equivalent option, at least not if `fnJC()` is used to provide the rate matrix.  Is there a natural way to employ this approach in RevBayes?

**Comment by wrightaprilm** [comment, marked as answer]
Hi Martin, 

We discuss this a bit in the [discrete tree estimation tutorial](https://revbayes.github.io/tutorials/morph_tree/). It's useable via the siteMatrices because you're functionally making a vector of Q-matrices that are compiled from different combinations of exchangeabilities and Pi values using the fnF81 function.  For binary characters, this should be what MrBayes is doing. You can manipulate both the number of categories and the shape of the Beta. Equivalency for the multistate characters are a bit murkier, as the explanation in the MrBayes manual is that these are fit via MCMC and the source code on this isn't greatly helpful.

It should be noted that ideally we'd probably be estimating a separate Pi per character because characters aren't equivalent. I know others are working on this, so perhaps someone else would like to hop in on that.

**Reply by ms609** [reply]
Oh, that's fantastic, thanks @wrightaprilm – just what I need.  And thanks for the additional explanation; this gives me a way in to understanding what RevBayes is doing mathematically.  So if I've read the tutorial code correctly, this allows the value of beta_scale to be inferred during MCMC – rather than repeating inference under a few different values as you did in your 2016 study?

**Reply by wrightaprilm** [reply]
Right - that's the idea. Drawing it from a distribution and optimizing via MCMC. I think @davidcerny had at one point written a script to compare the exact models I looked at in that paper via model averaging.

**Reply by davidcerny** [reply]
Yes indeed, here's my mess of a master script (a script-writing script, if you will) in case anyone finds it useful :p
[generate_rev_and_slurm_scripts.txt](https://github.com/user-attachments/files/17851294/generate_rev_and_slurm_scripts.txt)
